### PR TITLE
dhcp4client: a more powerful client.

### PR DESCRIPTION
**Pull request by hugelgupf**
here's some crazy code and tests to go with it.

### CODE REVIEW

File: dhcp4client/client.go [hunk 1]
```diff
@@ -78,77 +126,267 @@ func (c *Client) Close() error {
...
+// just have one dedicated goroutine for reading from the UDP socket, and use a
+// request and response queue.
+func (c *Client) SendAndRead(ctx context.Context, dest *net.UDPAddr, p *dhcp4.Packet) (<-chan *ClientPacket, <-chan *ClientError) {
+	out := make(chan *ClientPacket, 10)
+	errOut := make(chan *ClientError, 1)
+	go c.ParallelSendAndRead(ctx, dest, p, out, errOut)
+	return out, errOut
+}
+
+// ParallelSendAndRead sends the given packet `dest` to `to` and reads
```
Comment (GanShun) [inline dhcp4client/client.go]: sends the packet `dest` to `to`?

File: dhcp4client/client.go [hunk 2]
```diff
@@ -78,77 +126,267 @@ func (c *Client) Close() error {
...
+//
+// Any valid DHCP packet received with the correct Transaction ID is sent on
+// `out`.
+//
+// SendAndRead retries sending the packet and receiving responses according to
+// the configured number of c.retry, using a response timeout of c.timeout.
+//
+// TODO(hugelgupf): Make this a little state machine of packet types. See RFC
+// 2131, Section 4.4, Figure 5.
+func (c *Client) ParallelSendAndRead(ctx context.Context, dest *net.UDPAddr, p *dhcp4.Packet, out chan<- *ClientPacket, errCh chan<- *ClientError) {
```
Comment (GanShun) [inline dhcp4client/client.go]: This isn't really parallel..

File: dhcp4client/client.go [hunk 3]
```diff
@@ -78,77 +126,267 @@ func (c *Client) Close() error {
...
+				// No packets received. Sadness.
+				return timeoutCtx.Err()
+			default:
+			}
+
+			// Since a context can be canceled not just because of
+			// a deadline, we must check the context every once in
+			// a while. Use what is (hopefully) a small part of the
+			// context deadline rather than the context's deadline.
+			c.conn.SetReadDeadline(time.Now().Add(100 * time.Millisecond))
```
[the remark this answers is not in the document]
Comment (hugelgupf) [reply dhcp4client/client.go]: I think we can double back later and fix this. This is not ideal, but I don't really want to spend more time refactoring stuff.

File: dhcp4client/client.go [hunk 4]
```diff
@@ -78,77 +126,267 @@ func (c *Client) Close() error {
...
+	// This ensures that
+	// - we send at most one error on errCh; and
+	// - we don't forget to send err on errCh in the many return statements
+	//   of sendAndRead.
+	if err := c.sendAndRead(ctx, dest, p, out); err != nil {
+		errCh <- err
+	}
+}
+
+func (c *Client) sendAndRead(ctx context.Context, dest *net.UDPAddr, p *dhcp4.Packet, out chan<- *ClientPacket) *ClientError {
```
[the remark this answers is not in the document]
Comment (hugelgupf) [reply dhcp4client/client.go]: When they do almost the same thing, yes. The other function is just a wrapper around this to ensure I don't screw up the error channel.

**Pull request merged**